src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/IUpgradeContext.cs

```diff
@@ -27,6 +27,8 @@ public interface IUpgradeContext : IDisposable
 
         bool UpdateSolution(Solution updatedSolution);
 
+        string? BackupLocation { get; set; }
```

Comment:
Instead of an explicit property here, could you make it a property bag?

@mjrousos I know we've talked about this kind of thing in the past. Wanted to make sure you saw this conversation.

Reply:
@twsouthwick: Hopefully I understand correctly that with a property bag you mean a Dictionary construct. I  think that if we use a dictionary, the properties there should be used only from one context which would hold the property key. Otherwise the benefits gained would be outweighed by magic strings. BackupLocation is currently like that but if another step would want to use that, we would have to refactor this - and refactoring would be annoying because searching by reference would not work.

I was also thinking what kind of items would go into the property bag? Persisted items, settable items or something else? In both categories there will be items which are not strings and might require abstraction or generics.

I also noticed that IsComplete is seemingly designed to be persisted in file but it is never set. 

I'll happily make any changes, just verbalising my thoughts here.

Reply:
FWIW, I'm removing IsComplete in another unrelated PR. It's vestigial and can go away.

Reply:
Thanks for the thoughts! The reason I'd prefer more of a property bag construct is that it allows future steps to be able to add properties they care about without needing to add properties to the context and doesn't special case the BackupStep as something the context explicitly knows about.

For now, a prototype like the following could work well:

```csharp
public interface IUpgradeContext
{
  string GetPropertyValue(string propertyName);

  void SetPropertyValue(string propertyName, string value);
}
```

However, if that feels like a larger work item for you, we can punt that to a future time to get this feature in.

Reply:
As far as the `IsComplete` it was initially tracked in the state file, but was removed as it complicated some E2E experiences

Reply:
Hi,

I can look at this. I was thinking along the same lines but maybe with
```
public interface IUpgradeContext
{
  string GetPropertyValue(string propertyName);

  void SetPropertyValue(string propertyName, string value, bool persistent);
}
```
So that the context can specify whether the property is worth storing in the file.

Reply:
That's a reasonable API. Good idea!

Reply:
@twsouthwick Pushed the requested changes, can you take a look?